### DISCUSSION TITLE
Per-column JSON deserializers with psycopg2

### DISCUSSION THREAD

**Discussion by dev-zero** [sqlalchemy/sqlalchemy#6777] [Usage Questions] [answered]
Hi,

for a small project I am trying to find a way to set the JSON deserializer on a per-column basis, namely to be able to select whether JSON floats are returned as Python `Decimal` or as normal `float`.

For SQLite with `pysqlite` I had success with the following:

```python
from functools import partial

import wrapt
import simplejson
from sqlalchemy import TypeDecorator
from sqlalchemy.dialects import sqlite

class SimplejsonDialectWrapper(wrapt.ObjectProxy):
    """An engine wrapper with the JSON serializer set to simplejson with use_decimal=True"""
    def __init__(self, wrapped):
        super().__init__(wrapped)
        self._json_deserializer = partial(simplejson.loads, use_decimal=True)


class DecimalJSON(TypeDecorator):
    """Custom JSON column type which uses the SimplejsonDialectWrapper to to use simplejson.loads"""
    impl = sqlite.JSON

    def result_processor(self, dialect, coltype):
        return self.impl.result_processor(SimplejsonDialectWrapper(dialect), coltype)
```

which correctly gives me:

```
{'d': 3.14, 'f': 3.14}
{'d': Decimal('3.14'), 'f': Decimal('3.14')}
```

when called with the following model and code:

```python
engine = create_engine("sqlite+pysqlite:///:memory:", echo=True, future=True, json_serializer=simplejson.dumps)

Base = declarative_base()

class Data(Base):
    __tablename__ = "data"
    id = Column(Integer, primary_key=True)
    data1 = Column(JSON)
    data2 = Column(DecimalJSON)

Base.metadata.create_all(engine)

session = Session(engine)

data1 = Data(id=1, data1={"d": Decimal("3.14"), "f": 3.14}, data2={"d": Decimal("3.14"), "f": 3.14})
session.add(data1)
session.commit()

data1 = session.get(Data, 1)
print(data1.data1)
print(data1.data2)
```

Unfortunately the same solution does not seem to work with PostgreSQL (`impl` adapted to the PostgreSQL type, connecting to a PostgreSQL DB) since `psycopg2` does the serialization/deserialization directly. According to https://www.psycopg.org/docs/extras.html one can override it on a per-`execute` basis while keeping the default.
Is there a way to hook into that mechanism from SQLA? How?

Thanks a lot in advance,
Tiziano

**Comment by zzzeek** [comment, marked as answer]
the psycopg2 doc in question refers to the easiest way as just casting to string, so here's a solution that works on any backend agnostically; note that I couldn't get the "wrapt" thing to work so I just wrote this using the more standard type decorator API:

```py
from decimal import Decimal

import simplejson

from sqlalchemy import cast
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import String
from sqlalchemy import TypeDecorator
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session



class DecimalJSON(TypeDecorator):

    impl = JSON

    def column_expression(self, column):
        return cast(column, DecimalJSONStr)

    # I think this is likely not needed
    # def process_result_value(self, value, dialect):
    #     return simplejson.loads(value, use_decimal=True)


class DecimalJSONStr(TypeDecorator):
    impl = String

    def process_result_value(self, value, dialect):
        return simplejson.loads(value, use_decimal=True)


engine = create_engine(
# all URLs work
    "postgresql://scott:tiger@localhost/test",
#    "mysql://scott:tiger@localhost/test",
#    "sqlite://",
    echo=True,
    future=True,
    json_serializer=simplejson.dumps,
)

Base = declarative_base()


class Data(Base):
    __tablename__ = "data"
    id = Column(Integer, primary_key=True)
    data1 = Column(JSON)
    data2 = Column(DecimalJSON)


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

session = Session(engine)

data1 = Data(
    id=1,
    data1={"d": Decimal("3.14"), "f": 3.14},
    data2={"d": Decimal("3.14"), "f": 3.14},
)
session.add(data1)
session.commit()

data1 = session.get(Data, 1)
print(data1.data1)
print(data1.data2)
```

**Reply by dev-zero** [reply]
@zzzeek wow, thanks a lot, it works.

Just to check whether I understood it correctly: to avoid the JSON parsing by psycopg2 this introduces a `CAST` to string (`CAST(data.data2 AS VARCHAR)`) at the SQL level, by using a decorated string column type at the Python level, where the decoration overrides the default string loading (which now happens independently of the database library due to the string conversion) with a `simplejson.loads` to actually return the parsed dictionary?

Is there anything I have to be aware of with this solution (state tracking maybe)?
If I wanted to use a schema-based loading (with something like FastAPI), placing this call in the `process_result_value` you provided would be the correct place as well, right?

**Reply by zzzeek** [reply]
> @zzzeek wow, thanks a lot, it works.
> 
> Just to check whether I understood it correctly: to avoid the JSON parsing by psycopg2 this introduces a `CAST` to string (`CAST(data.data2 AS VARCHAR)`) at the SQL level, by using a decorated string column type at the Python level, where the decoration overrides the default string loading (which now happens independently of the database library due to the string conversion) with a `simplejson.loads` to actually return the parsed dictionary?

yes


> 
> Is there anything I have to be aware of with this solution (state tracking maybe)?

state tracking if you mean ORM stuff all occurs within Python using the in-python value of things, so for JSON since that's a mutable structure, the ORM by default won't detect if you change the structure in place.  if you need that tracking you'd want to use the [mutabledict](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html?highlight=mutabledict#sqlalchemy.ext.mutable.MutableDict) for that.   it should work w/ the above but if it doesnt then check back here. 

> If I wanted to use a schema-based loading (with something like FastAPI), placing this call in the `process_result_value` you provided would be the correct place as well, right?

I dont really know how FastAPI impacts what we're doing here, the term "schema-based loading" is not ringing a bell?  can you be more specific?

**Reply by dev-zero** [reply]
> state tracking if you mean ORM stuff all occurs within Python using the in-python value of things, so for JSON since that's a mutable structure, the ORM by default won't detect if you change the structure in place. if you need that tracking you'd want to use the [mutabledict](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html?highlight=mutabledict#sqlalchemy.ext.mutable.MutableDict) for that. it should work w/ the above but if it doesnt then check back here.

Ok, clear, thanks.

> > If I wanted to use a schema-based loading (with something like FastAPI), placing this call in the `process_result_value` you provided would be the correct place as well, right?
> 
> I dont really know how FastAPI impacts what we're doing here, the term "schema-based loading" is not ringing a bell? can you be more specific?

With the JSON loads/dumps we only have the option to specify based on (JSON) datatype how the parsed value should be represented in Python (which in the case of a JSON float could affected the value). This implies that will never get a mix of `Decimal` or `Float` back (always one or the other), even worse for a `Decimal` which got serialized to a JSON `int`, which after the deserialization will be an `int` no matter what. And the only way around this would be to instead use a JSON parser which always loads numerical values as strings and then apply a conversion to int/float/Decimal/numpy.float/whatever based on a schema. FastAPI seemed to be one of the few libraries handling this out-of-the-box as part of its schema handling, but after some more careful digging it seems they also just rely on Pydantic+orjson for it (which doesn't even support that use case yet).

**Reply by zzzeek** [reply]
by "schema" you mean JSON schema?   that would be all in-python stuff that interacts with SQLAlchemy as above.
